File: src/daemon/session.ts
import { randomBytes } from 'crypto';

/** A browser session expires after this much idle time. The vault does not. */
export const SESSION_IDLE_MS = 30 * 60 * 1000;
const COOKIE_NAME = 'agentio_session';

const sessions = new Map<string, number>(); // id -> lastSeen

export function createSession(now = Date.now()): string {
  const id = randomBytes(32).toString('base64url');
  sessions.set(id, now);
  return id;
}

/** True when the request carries a live session; touching it extends the idle window. */
export function hasSession(request: Request, now = Date.now()): boolean {
  const id = sessionIdFrom(request);
  if (!id) return false;
  const lastSeen = sessions.get(id);
  if (lastSeen === undefined) return false;
  if (now - lastSeen > SESSION_IDLE_MS) {
    sessions.delete(id);
    return false;
  }
  sessions.set(id, now);
  return true;
}

/** Forget the session this request carries, if any. Other browsers stay signed in. */
export function deleteSession(request: Request): void {
  const id = sessionIdFrom(request);
  if (id) sessions.delete(id);
}

export function clearSessions(): void {
  sessions.clear();
}

/**
 * `Secure` only when the request itself came over HTTPS: browsers accept a
 * Secure cookie over plain HTTP for localhost and 127.0.0.1 but drop it for any
 * other host, which would make an unlock over http://0.0.0.0 or a LAN address
 * silently fail. Behind the TLS proxy the forwarded protocol says https.
 */
export function isSecureRequest(request: Request): boolean {
  return request.headers.get('x-forwarded-proto') === 'https' || new URL(request.url).protocol === 'https:';
}

const cookieAttributes = (secure: boolean) => `HttpOnly; ${secure ? 'Secure; ' : ''}SameSite=Strict; Path=/`;

export function sessionCookie(id: string, secure: boolean): string {
  return `${COOKIE_NAME}=${id}; ${cookieAttributes(secure)}`;
}

export function expiredSessionCookie(secure: boolean): string {
  return `${COOKIE_NAME}=; ${cookieAttributes(secure)}; Max-Age=0`;
}

function sessionIdFrom(request: Request): string | null {
  const header = request.headers.get('cookie');
  if (!header) return null;
  for (const part of header.split(';')) {
    const [name, ...rest] = part.trim().split('=');
    if (name === COOKIE_NAME) return rest.join('=') || null;
  }
  return null;
}

File: go/internal/daemon/session.go
package daemon

import (
	"crypto/rand"
	"encoding/base64"
	"net/http"
	"strings"
	"sync"
	"time"
)

const (
	sessionIdle = 30 * time.Minute
	cookieName  = "agentio_session"
)

var (
	sessionMu sync.Mutex
	sessions  = map[string]time.Time{}
)

func ResetSessions() {
	sessionMu.Lock()
	sessions = map[string]time.Time{}
	sessionMu.Unlock()
}

func CreateSession(now time.Time) string {
	b := make([]byte, 32)
	_, _ = rand.Read(b)
	id := base64.RawURLEncoding.EncodeToString(b)
	sessionMu.Lock()
	sessions[id] = now
	sessionMu.Unlock()
	return id
}

func HasSession(r *http.Request, now time.Time) bool {
	id := sessionID(r)
	if id == "" {
		return false
	}
	sessionMu.Lock()
	defer sessionMu.Unlock()
	last, ok := sessions[id]
	if !ok {
		return false
	}
	if now.Sub(last) > sessionIdle {
		delete(sessions, id)
		return false
	}
	sessions[id] = now
	return true
}

func DeleteSession(r *http.Request) {
	id := sessionID(r)
	if id == "" {
		return
	}
	sessionMu.Lock()
	delete(sessions, id)
	sessionMu.Unlock()
}

func ClearSessions() {
	ResetSessions()
}

// SecureRequest is whether the Set-Cookie gets Secure: the request came over
// TLS, or the fronting proxy says it did with exactly "https", as Bun compares.
func SecureRequest(r *http.Request) bool {
	return strings.Join(r.Header.Values("X-Forwarded-Proto"), ", ") == "https" || r.TLS != nil
}

func SessionCookie(id string, secure bool) string {
	return cookieName + "=" + id + "; " + cookieAttrs(secure)
}

func ExpiredSessionCookie(secure bool) string {
	return cookieName + "=; " + cookieAttrs(secure) + "; Max-Age=0"
}

func cookieAttrs(secure bool) string {
	s := "HttpOnly; "
	if secure {
		s += "Secure; "
	}
	return s + "SameSite=Strict; Path=/"
}

func sessionID(r *http.Request) string {
	header := r.Header.Get("Cookie")
	if header == "" {
		return ""
	}
	for _, part := range strings.Split(header, ";") {
		part = strings.TrimSpace(part)
		name, value, ok := strings.Cut(part, "=")
		if ok && name == cookieName {
			return value
		}
	}
	return ""
}
